### PR TITLE
2244 fix interval trigger

### LINKED ISSUE #2244: IntervalTrigger may become True at the improper timing when resuming learning
For example,
`IntervalTrigger(3, 'epoch')` will return True at the initial iteration when resuming from epoch 5.
This is probably because 
https://github.com/pfnet/chainer/blob/master/chainer/training/triggers/interval.py#L26
is not taken over from snapshot.

### PR DESCRIPTION
**Pull request by Hakuyume**
Fix #2244 
This code assumes that the batchsize does not change during training.

### CODE REVIEW

File: tests/chainer_tests/training_tests/triggers_tests/test_minmax_trigger.py [hunk 1]
```diff
@@ -26,6 +26,10 @@ def epoch(self):
         return 1
 
     @property
+    def epoch_detail(self):
```
Comment (delta2323) [inline tests/chainer_tests/training_tests/triggers_tests/test_minmax_trigger.py]: I am wondering why we modify the unit test of `MinMaxTrigger` as this PR is about `IntervalTrigger`?

Comment (Hakuyume) [inline tests/chainer_tests/training_tests/triggers_tests/test_minmax_trigger.py]: This is because `MinMaxTrigger` calls `IntervalTrigger` internally.
https://github.com/pfnet/chainer/blob/master/chainer/training/triggers/minmax_value_trigger.py#L24
https://github.com/pfnet/chainer/blob/master/chainer/training/util.py#L39

Without this modification, the tests of `MinMaxTrigger` failled.

Comment (Hakuyume) [inline tests/chainer_tests/training_tests/triggers_tests/test_minmax_trigger.py]: Sorry, I could pass the tests of `MinMaxTrigger` without modifying the test code.

**Comment by delta2323** [comment]
> This code assumes that the batchsize does not change during training.

Could you add the assumption to the document of `IntervalTrigger`?

**Comment by Hakuyume** [comment]
I added note.

**Comment by delta2323** [comment]
Could you run `examples/mnist/train_mnist.py` with `--epoch=1`? Our example test failed with division-by-zero error.

```
15:13:46 + coverage run -a --branch examples/mnist/train_mnist.py --epoch=1 --unit=10
15:13:46 Running mnist example
15:16:51 Traceback (most recent call last):
15:16:51   File "examples/mnist/train_mnist.py", line 115, in <module>
15:16:51     main()
15:16:51   File "examples/mnist/train_mnist.py", line 112, in main
15:16:51     trainer.run()
15:16:51   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 292, in run
15:16:51     while not stop_trigger(self):
15:16:51   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/triggers/interval.py", line 50, in __call__
15:16:51     prev = (epoch * (iteration - 1) / iteration) // self.period
15:16:51 ZeroDivisionError: float division by zero
```

**Comment by Hakuyume** [comment]
I missed the case that `IntervalTrigger` is used as stop trigger.
I added some test codes and fixed `IntervalTrigger`.

**Comment by delta2323** [comment]
I am a little worried ()
I come to think

**Comment by delta2323** [comment]
I am a little concern about the inconsistent behavior caused by the rounding error. For example, I have such a case in my mind: suppose that mathematically `updater.epoch_detail * (iteration - 1) / iteration` equals to some integer (say 1000) exactly, but because of the rounding problem, the value could be less than the integer. Then, the actual `prev` value could be different from expected one.

But as I have not checked it yet, It could be unnecessary worry. Do you have an opinion on this problem?

**Comment by Hakuyume** [comment]
As you pointed out, the rounding error causes an undesired behavior.
If `(updater.epoch_detail * (iteration - 1) / iteration) // period` is less than `prev // period`, this trigger fires twice (before and after resuming).
I could easily reproduce this behavior using `round()` in `DummyUpdater.epoch_detail`.

To fix this bug, I think we have to modify `Updater` to save the previous epoch in the snapshot.

**Comment by delta2323** [comment]
Thank you for considering the problem. It seems reasonable to me to have `Updater` serialize the value of previous `epoch_detail` is one of the option. 

The other choice I come up with is to implement `serialize` method to `IntervalTrigger` to save `prev` value and to call the method in `Updater.serialize`. But the former solution seems better to me because (1) `Updater` have to check if the trigger is `IntervalTrigger` or has the `serialize` method (2) it is natural for `IntervalTrigger` to get the previous value of `epoch_detail` as well as current `epoch_detail`.

In either case, we must be careful to keep the [model format compatibility](http://docs.chainer.org/en/latest/compatibility.html#model-format-compatibility).

**Comment by Hakuyume** [comment]
Thank you for your suggestion.
I also think the former solution is better because all resumable triggers need the previous value of `epoch_detail`.
It is redundant that each trigger keeps its own value of previous `epoch_detail`.

I will implement the former solution.

**Comment by Hakuyume** [comment]
I updated `SerialIterator`, `MultiProcessIterator`, and `StandardUpdater` to support `previous_epoch_detail`.
For the model format compatibility, an iterator guesses `previous_epoch_detail` if its information is absent.

File: chainer/training/triggers/interval.py [hunk 2]
```diff
@@ -40,9 +39,11 @@ def __call__(self, trainer):
         """
         updater = trainer.updater
         if self.unit == 'epoch':
-            prev = self.count
-            self.count = updater.epoch_detail // self.period
-            return prev != self.count
+            epoch = updater.epoch_detail
+            previous_epoch = updater.previous_epoch_detail
```
Comment (delta2323) [inline chainer/training/triggers/interval.py]: I think `getattr(updater, previous_epoch_detail, None)` is safer.

File: chainer/training/triggers/interval.py [hunk 3]
```diff
@@ -40,9 +39,11 @@ def __call__(self, trainer):
         """
         updater = trainer.updater
         if self.unit == 'epoch':
-            prev = self.count
-            self.count = updater.epoch_detail // self.period
-            return prev != self.count
+            epoch = updater.epoch_detail
```
Comment (delta2323) [inline chainer/training/triggers/interval.py]: It is confusing as Chainer clearly distinguish `epoch` with `epoch_detail`.

File: tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py [hunk 4]
```diff
@@ -185,4 +244,94 @@ def test_copy_not_repeat(self):
         for _ in range(2):
             self.assertRaises(StopIteration, copy_it.next)
 
+
+@testing.parameterize(*testing.product({
+    'n_prefetch': [1, 2],
+    'shared_mem': [None, 1000000],
+}))
+class TestMultiprocessIteratorSerialize(unittest.TestCase):
```
Comment (delta2323) [inline tests/chainer_tests/iterators_tests/test_multiprocess_iterator.py]: Is it possible to send this unittest as another PR?

File: tests/chainer_tests/iterators_tests/test_serial_iterator.py [hunk 5]
```diff
@@ -147,4 +211,83 @@ def test_iterator_shuffle_nondivisible(self):
         self.assertNotEqual(out[0:10], out[10:20])
 
 
+class TestSerialIteratorSerialize(unittest.TestCase):
```
Comment (delta2323) [inline tests/chainer_tests/iterators_tests/test_serial_iterator.py]: ditto (could you extract it and make another PR?)

**Comment by Hakuyume** [comment]
I created another PR for serialization tests #2346.

**Comment by Hakuyume** [comment]
I also divide `previous_epoch_detail` feature into another PR #2347.

File: chainer/training/triggers/interval.py (hunk 2 above)
Comment (Hakuyume) [inline chainer/training/triggers/interval.py]: `getattr` is safer in the terms of not raising any exception.
But setting `previous_epoch(_detail)` to `None` causes undesired behavior (same as #2244).
I think it is better to raise exception (or warning) when `updater` does not have `previous_epoch_detail` property.

File: chainer/training/triggers/interval.py (hunk 3 above)
Comment (Hakuyume) [inline chainer/training/triggers/interval.py]: I will fix it.

**Comment by delta2323** [comment]
I will review this PR after #2347 is done.

**Comment by Hakuyume** [comment]
This PR has a lot of unused commits because I changed the direction of this PR. I am planning to rebase this PR on #2347 and #2472. If it is better to keep all history, I will merge instead.

**Comment by delta2323** [comment]
@Hakuyume It is OK to close this PR and create another PR on top of them.

**Comment by Hakuyume** [comment]
@delta2323 Thank you. I made another PR (#2484).
